3party/opening_hours/osm_time_range.hpp

```diff
@@ -27,108 +27,491 @@
+    eDusk
+  };
 
-    TimeSpan() : flags(Time::eNone) {}
+  using THours = std::chrono::hours;
+  using TMinutes = std::chrono::minutes;
 
-    friend std::ostream & operator << (std::ostream & s, TimeSpan const & span);
-  };
+ public:
+  Time() = default;
+  Time(Time const &) = default;
+  Time(THours const hours);
+  Time(TMinutes const minutes);
+  Time(THours const hours, TMinutes const minutes);
+  Time(EEvent const event);
 
-  class Weekdays
-  {
-  public:
-    uint8_t weekdays;
-    uint16_t nth;
-    int32_t offset;
+  Time & operator=(Time const &) = default;
 
-    Weekdays() : weekdays(0), nth(0), offset(0) {}
+  THours::rep GetHoursCount() const;
+  TMinutes::rep GetMinutesCount() const;
 
-    friend std::ostream & operator << (std::ostream & s, Weekdays const & w);
-  };
+  THours GetHours() const;
+  TMinutes GetMinutes() const;
+
+  void SetHours(THours const hours);
+  void SetMinutes(TMinutes const minutes);
+
+  EEvent GetEvent() const {return m_event;}
+  void SetEvent(EEvent const event);
+
+  bool IsEvent() const;
+  bool IsEventOffset() const;
+  bool IsHoursMinutes() const;
+  bool IsMinutes() const;
+  bool IsTime() const;
+  bool HasValue() const;
+
+  Time & operator-(Time const & time);
+  Time & operator-();
+
+ private:
+  Time GetEventTime() const;
+
+ private:
+  EEvent m_event{EEvent::eNotEvent};
+  TMinutes m_duration{TMinutes::zero()};
+  TEStateRep m_state{EState::eIsNotTime};
+};
+
+inline constexpr Time::THours operator ""_h(unsigned long long h)
+{
+  return Time::THours(h);
+}
+
+inline constexpr Time::TMinutes operator ""_min(unsigned long long m)
+{
+  return Time::TMinutes(m);
+}
+
+std::ostream & operator<<(std::ostream & ost, Time::EEvent const event);
+std::ostream & operator<<(std::ostream & ost, Time const & time);
+
+class Timespan
+{
+ public:
+  Timespan() = default;
+  Timespan(Timespan const &) = default;
+  Timespan(Time const & start, bool plus = false);
+  Timespan(Time const & start, Time const & end, bool plus = false);
+  Timespan(Time const & start, Time const & end, Time const & period);
+
+  bool IsOpen() const;
+  bool HasPlus() const;
+  bool HasPeriod() const;
+
+  Time const & GetStart() const;
+  Time const & GetEnd() const;
+  Time const & GetPeriod() const;
+
+  void SetStart(Time const & start);
+  void SetEnd(Time const & end);
+  void SetPeriod(Time const & period);
+  void SetPlus(bool const plus);
+
+  bool IsValid() const;
+
+ private:
+  Time m_start;
+  Time m_end;
+  Time m_period;
+  bool m_plus{false};
+};
+
+using TTimespans = std::vector<Timespan>;
 
-  class State
+std::ostream & operator<<(std::ostream & ost, Timespan const & span);
+std::ostream & operator<<(std::ostream & ost, osmoh::TTimespans const & timespans);
+
+class NthEntry
+{
+ public:
+  enum class ENth
   {
-  public:
-    enum EState {
-      eUnknown = 0,
-      eClosed = 1,
-      eOpen = 2
-    };
+    None,
+    First,
+    Second,
+    Third,
+    Fourth,
+    Fifth
+  };
+
+ public:
```

Comment:
No need in such duplication of access specifiers. Usually we declare public fields and methods first, then protected and finally private.